File: vs-qdrant/python/official_scenarios.py
"""Official Qdrant SDK (qdrant-client) scenario implementations.

One method per benchmark scenario. Kept in its own file so the LOC report
counts exactly the application code a Qdrant user writes — no harness,
timing, or parity logic is included.

Every method returns plain Python objects (hits, counts, dicts) so the
harness can compare results against the QQL side without knowing either
SDK's response shapes.
"""
from __future__ import annotations

import time

import numpy as np
from qdrant_client import QdrantClient
from qdrant_client import models as M

from config import BATCH_BERLIN, BATCH_LEGAL


class OfficialScenarios:
    """qdrant-client 1.19 — the reference Qdrant Python SDK."""

    def __init__(self, url: str, *, prefer_grpc: bool = False):
        self.client = QdrantClient(url=url, timeout=120, check_compatibility=False,
                                   prefer_grpc=prefer_grpc)

    def close(self) -> None:
        self.client.close()

    # ------------------------------------------------------------ setup ----
    def drop_collection(self, name: str) -> None:
        self.client.delete_collection(name)

    def create_berlin(self, name: str) -> None:
        self.client.create_collection(
            collection_name=name,
            vectors_config={
                "dense": M.VectorParams(size=384, distance=M.Distance.COSINE),
            },
            sparse_vectors_config={
                # BM25 vectors need the idf modifier for true BM25 scoring —
                # the same default QQL applies to `sparse SPARSE` columns.
                "bm25": M.SparseVectorParams(modifier=M.Modifier.IDF)
            },
        )
        self.client.create_payload_index(name, "district", M.PayloadSchemaType.KEYWORD)
        self.client.create_payload_index(name, "price", M.PayloadSchemaType.FLOAT)

    def create_legal(self, name: str) -> None:
        self.client.create_collection(
            collection_name=name,
            vectors_config={
                "dense": M.VectorParams(size=384, distance=M.Distance.COSINE),
                "colbert": M.VectorParams(
                    size=128,
                    distance=M.Distance.COSINE,
                    multivector_config=M.MultiVectorConfig(
                        comparator=M.MultiVectorComparator.MAX_SIM
                    ),
                    hnsw_config=M.HnswConfigDiff(m=0),
                ),
            },
            sparse_vectors_config={
                # BM25 vectors need the idf modifier for true BM25 scoring —
                # the same default QQL applies to `sparse SPARSE` columns.
                "bm25": M.SparseVectorParams(modifier=M.Modifier.IDF)
            },
        )
        self.client.create_payload_index(name, "court", M.PayloadSchemaType.KEYWORD)
        self.client.create_payload_index(name, "year", M.PayloadSchemaType.INTEGER)

    # ----------------------------------------------------------- ingest ----
    def ingest_berlin(self, name: str, docs, dense, sparse) -> float:
        dense_list = dense.tolist() if hasattr(dense, "tolist") else dense
        sparse_vecs = [
            M.SparseVector(indices=s["indices"], values=s["values"])
            for s in sparse
        ]
        payloads = [{k: v for k, v in d.items() if k != "id"} for d in docs]
        ids = [d["id"] for d in docs]
        t0 = time.perf_counter()
        for i in range(0, len(docs), BATCH_BERLIN):
            self.client.upsert(
                name,
                points=M.Batch(
                    ids=ids[i : i + BATCH_BERLIN],
                    vectors={
                        "dense": dense_list[i : i + BATCH_BERLIN],
                        "bm25": sparse_vecs[i : i + BATCH_BERLIN],
                    },
                    payloads=payloads[i : i + BATCH_BERLIN],
                ),
                wait=False,
            )
        return time.perf_counter() - t0

    def ingest_legal(self, name: str, docs, dense, sparse, colbert) -> float:
        dense_list = dense.tolist() if hasattr(dense, "tolist") else dense
        sparse_vecs = [
            M.SparseVector(indices=s["indices"], values=s["values"])
            for s in sparse
        ]
        payloads = [{k: v for k, v in d.items() if k != "id"} for d in docs]
        ids = [d["id"] for d in docs]
        t0 = time.perf_counter()
        for i in range(0, len(docs), BATCH_LEGAL):
            self.client.upsert(
                name,
                points=M.Batch(
                    ids=ids[i : i + BATCH_LEGAL],
                    vectors={
                        "dense": dense_list[i : i + BATCH_LEGAL],
                        "bm25": sparse_vecs[i : i + BATCH_LEGAL],
                        "colbert": colbert[i : i + BATCH_LEGAL],
                    },
                    payloads=payloads[i : i + BATCH_LEGAL],
                ),
                wait=False,
            )
        return time.perf_counter() - t0

    # ------------------------------------------------------------- reads ----
    def query_dense(self, name: str, qvec: list[float]) -> list:
        return self.client.query_points(
            name, query=qvec, using="dense", limit=10, with_payload=True,
        ).points

    def query_dense_filtered(self, name: str, qvec: list[float]) -> list:
        flt = M.Filter(must=[
            M.FieldCondition(key="price", range=M.Range(lt=150.0)),
            M.FieldCondition(key="guests", range=M.Range(gte=2)),
        ])
        return self.client.query_points(
            name, query=qvec, using="dense", limit=10, query_filter=flt, with_payload=True,
        ).points

    def query_sparse(self, name: str, svec: dict) -> list:
        return self.client.query_points(
            name,
            query=M.SparseVector(indices=svec["indices"], values=svec["values"]),
            using="bm25", limit=10, with_payload=True,
        ).points

    def query_hybrid(self, name: str, qvec: list[float], svec: dict) -> list:
        # hnsw_ef=128 on the dense leg: fused rankings must be deterministic
        # across the two independently built collections.
        return self.client.query_points(
            name,
            prefetch=[
                M.Prefetch(query=qvec, using="dense", limit=50,
                           params=M.SearchParams(hnsw_ef=128, exact=False)),
                M.Prefetch(
                    query=M.SparseVector(indices=svec["indices"], values=svec["values"]),
                    using="bm25", limit=50,
                ),
            ],
            query=M.FusionQuery(fusion=M.Fusion.RRF),
            limit=10, with_payload=True,
        ).points

    def query_colbert(self, name: str, mvec: list) -> list:
        return self.client.query_points(
            name, query=mvec, using="colbert", limit=10, with_payload=True,
        ).points

    def scroll_pages(self, name: str, pages: int, batch: int) -> list[int]:
        ids, offset = [], None
        for _ in range(pages):
            points, offset = self.client.scroll(
                name, limit=batch, offset=offset, with_payload=True,
            )
            ids.extend(p.id for p in points)
            if offset is None:
                break
        return ids

    def count_berlin(self, name: str) -> int:
        flt = M.Filter(must=[M.FieldCondition(key="price", range=M.Range(lt=150.0))])
        return self.client.count(name, count_filter=flt, exact=True).count

    def count_legal(self, name: str) -> int:
        flt = M.Filter(must=[M.FieldCondition(key="year", range=M.Range(gte=2010))])
        return self.client.count(name, count_filter=flt, exact=True).count

    def facet_district(self, name: str) -> dict:
        hits = self.client.facet(name, key="district", limit=20, exact=True).hits
        return {h.value: h.count for h in hits}

    # ----------------------------------------------------------- writes ----
    def update_payload(self, name: str) -> None:
        flt = M.Filter(must=[M.FieldCondition(key="district", match=M.MatchValue(value="Mitte"))])
        self.client.set_payload(name, payload={"rating": 4.5}, points=flt, wait=True)

    def delete_by_filter(self, name: str) -> None:
        flt = M.Filter(must=[M.FieldCondition(key="price", range=M.Range(gt=250.0))])
        self.client.delete(name, points_selector=M.FilterSelector(filter=flt), wait=True)

    def prepared_rerun(self, name: str, qvecs: list[list[float]]) -> list:
        """No prepared statements in the official SDK: repeat the call."""
        hits = []
        for v in qvecs:
            hits = self.client.query_points(
                name, query=v, using="dense", limit=10, with_payload=True,
            ).points
        return hits

File: vs-qdrant/node/official_scenarios.js
/**
 * Official Qdrant SDK (@qdrant/js-client-rest) scenario implementations.
 *
 * One method per benchmark scenario — the exact application code a Qdrant
 * JS user writes. Symmetric with qql_scenarios.js; the harness measures both.
 */
const { QdrantClient } = require("@qdrant/js-client-rest");

const BATCH_BERLIN = 100;
const BATCH_LEGAL = 32;

class OfficialScenarios {
  constructor(url) {
    this.client = new QdrantClient({ url, checkCompatibility: false, timeout: 120_000 });
  }

  async close() {
    // js-client has no explicit close
  }

  // ------------------------------------------------------------ setup ----
  async dropCollection(name) {
    try { await this.client.deleteCollection(name); } catch (_) { /* absent */ }
  }

  async createBerlin(name) {
    await this.client.createCollection(name, {
      vectors: { dense: { size: 384, distance: "Cosine" } },
      sparse_vectors: { bm25: { modifier: "idf" } },
    });
    await this.client.createPayloadIndex(name, { field_name: "district", field_schema: "keyword" });
    await this.client.createPayloadIndex(name, { field_name: "price", field_schema: "float" });
  }

  async createLegal(name) {
    await this.client.createCollection(name, {
      vectors: {
        dense: { size: 384, distance: "Cosine" },
        colbert: {
          size: 128, distance: "Cosine",
          multivector_config: { comparator: "max_sim" },
          hnsw_config: { m: 0 },
        },
      },
      sparse_vectors: { bm25: { modifier: "idf" } },
    });
    await this.client.createPayloadIndex(name, { field_name: "court", field_schema: "keyword" });
    await this.client.createPayloadIndex(name, { field_name: "year", field_schema: "integer" });
  }

  // ----------------------------------------------------------- ingest ----
  _points(docs, dense, sparse, colbert) {
    return docs.map((doc, i) => {
      const vector = {
        dense: Array.from(dense.subarray(i * 384, (i + 1) * 384)),
        bm25: { indices: sparse[i].indices, values: sparse[i].values },
      };
      if (colbert) vector.colbert = colbert[i];
      const { id, ...payload } = doc;
      return { id, vector, payload };
    });
  }

  async ingestBerlin(name, docs, dense, sparse) {
    const points = this._points(docs, dense, sparse, null);
    const t0 = process.hrtime.bigint();
    for (let i = 0; i < points.length; i += BATCH_BERLIN) {
      await this.client.upsert(name, { points: points.slice(i, i + BATCH_BERLIN), wait: false });
    }
    return Number(process.hrtime.bigint() - t0) / 1e9;
  }

  async ingestLegal(name, docs, dense, sparse, colbert) {
    const points = this._points(docs, dense, sparse, colbert);
    const t0 = process.hrtime.bigint();
    for (let i = 0; i < points.length; i += BATCH_LEGAL) {
      await this.client.upsert(name, { points: points.slice(i, i + BATCH_LEGAL), wait: false });
    }
    return Number(process.hrtime.bigint() - t0) / 1e9;
  }

  // ------------------------------------------------------------- reads ----
  async queryDense(name, qvec) {
    return (await this.client.query(name, {
      query: qvec, using: "dense", limit: 10, with_payload: true,
    })).points;
  }

  async queryDenseFiltered(name, qvec) {
    return (await this.client.query(name, {
      query: qvec, using: "dense", limit: 10, with_payload: true,
      filter: {
        must: [
          { key: "price", range: { lt: 150.0 } },
          { key: "guests", range: { gte: 2 } },
        ],
      },
    })).points;
  }

  async querySparse(name, svec) {
    return (await this.client.query(name, {
      query: { indices: svec.indices, values: svec.values },
      using: "bm25", limit: 10, with_payload: true,
    })).points;
  }

  async queryHybrid(name, qvec, svec) {
    // hnsw_ef=128 on the dense leg: fused rankings must be deterministic
    // across the two independently built collections.
    return (await this.client.query(name, {
      prefetch: [
        { query: qvec, using: "dense", limit: 50, params: { hnsw_ef: 128 } },
        { query: { indices: svec.indices, values: svec.values }, using: "bm25", limit: 50 },
      ],
      query: { fusion: "rrf" },
      limit: 10, with_payload: true,
    })).points;
  }

  async queryColbert(name, mvec) {
    return (await this.client.query(name, {
      query: mvec, using: "colbert", limit: 10, with_payload: true,
    })).points;
  }

  async scrollPages(name, pages, batch) {
    const ids = [];
    let offset;
    for (let p = 0; p < pages; p++) {
      const { points, next_page_offset: next } = await this.client.scroll(name, {
        limit: batch, offset, with_payload: true,
      });
      for (const pt of points) ids.push(pt.id);
      offset = next;
      if (offset === undefined || offset === null) break;
    }
    return ids;
  }

  async countBerlin(name) {
    return (await this.client.count(name, {
      filter: { must: [{ key: "price", range: { lt: 150.0 } }] }, exact: true,
    })).count;
  }

  async countLegal(name) {
    return (await this.client.count(name, {
      filter: { must: [{ key: "year", range: { gte: 2010 } }] }, exact: true,
    })).count;
  }

  async facetDistrict(name) {
    const res = await this.client.facet(name, { key: "district", limit: 20, exact: true });
    const map = {};
    for (const h of res.hits) map[h.value] = h.count;
    return map;
  }

  // ----------------------------------------------------------- writes ----
  async updatePayload(name) {
    await this.client.setPayload(name, {
      payload: { rating: 4.5 },
      filter: { must: [{ key: "district", match: { value: "Mitte" } }] },
      wait: true,
    });
  }

  async deleteByFilter(name) {
    await this.client.delete(name, {
      filter: { must: [{ key: "price", range: { gt: 250.0 } }] },
      wait: true,
    });
  }

  async preparedRerun(name, qvecs) {
    // No prepared statements in the official SDK: repeat the call.
    let hits;
    for (const v of qvecs) {
      hits = (await this.client.query(name, {
        query: v, using: "dense", limit: 10, with_payload: true,
      })).points;
    }
    return hits;
  }
}

module.exports = { OfficialScenarios };
